receiver/mongodbreceiver/client.go

```diff
@@ -0,0 +1,222 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.uber.org/zap"
+)
+
+// Client is an interface that exposes functionality towards a mongo environment
+type Client interface {
+	ListDatabaseNames(ctx context.Context, filters interface{}, opts ...*options.ListDatabasesOptions) ([]string, error)
+	Disconnect(context.Context) error
+	Connect(context.Context) error
+	TestConnection(context.Context) (*mongoTestConnectionResponse, error)
+	ServerStatus(ctx context.Context, DBName string) (bson.M, error)
+	DBStats(ctx context.Context, DBName string) (bson.M, error)
+}
+
+// driver is the underlying mongo.Client that Client invokes to actually do actions
+// against the mongo instance
+type driver interface {
+	Database(dbName string, opts ...*options.DatabaseOptions) *mongo.Database
+	ListDatabaseNames(ctx context.Context, filters interface{}, opts ...*options.ListDatabasesOptions) ([]string, error)
+	Connect(context.Context) error
+	Disconnect(context.Context) error
+	Ping(ctx context.Context, rp *readpref.ReadPref) error
+}
+
+// mongodbClient is a mongodb metric scraper client
+type mongodbClient struct {
+	// underlying mongo client driver
+	driver         driver
+	username       string
+	password       string
+	hosts          []string
+	replicaSetName string
+	logger         *zap.Logger
+	timeout        time.Duration
+	tlsConfig      *tls.Config
+}
+
+// NewClient creates a new client to connect and query mongo for the
+// mongodbreceiver
+func NewClient(config *Config, logger *zap.Logger) (Client, error) {
+	tlsConfig, err := config.LoadTLSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("invalid tls configuration: %w", err)
+	}
+	hosts := []string{}
+	for _, ep := range config.Hosts {
+		hosts = append(hosts, ep.Endpoint)
+	}
+
+	return &mongodbClient{
+		hosts:          hosts,
+		username:       config.Username,
+		password:       config.Password,
+		replicaSetName: config.ReplicaSet,
+		timeout:        config.Timeout,
+		logger:         logger,
+		tlsConfig:      tlsConfig,
+	}, nil
+}
+
+// Connect establishes a connection to mongodb instance
+func (c *mongodbClient) Connect(ctx context.Context) error {
+	if err := c.ensureClient(ctx); err != nil {
+		return fmt.Errorf("unable to instantiate mongo client: %w", err)
+	}
+	return nil
+}
+
+// Disconnect closes attempts to close any open connections
+func (c *mongodbClient) Disconnect(ctx context.Context) error {
+	if c.isConnected(ctx) {
```

Comment:
There's an example in the docs [here](https://pkg.go.dev/go.mongodb.org/mongo-driver/mongo#example-Connect-Ping) that shows `Disconnect` being called w/o regard for whether `Connect` succeeded. If we don't need this check, then `isConnected` is only used once, and can also be rolled up into `Connect`. This would normalize the code a lot in my opinion.